rust/otap-dataflow/crates/otlp/src/grpc.rs

```diff
@@ -155,3 +155,89 @@ pub enum OTLPData {
     /// Profiles Object
     Profiles(ExportProfilesServiceRequest),
 }
+
+impl otap_df_traits::Retryable for OTLPData {
+    fn id(&self) -> u64 {
+        use prost::Message;
+        use std::collections::hash_map::DefaultHasher;
+        use std::hash::{Hash, Hasher};
+
+        let mut hasher = DefaultHasher::new();
+
+        // Include enum variant discriminant for uniqueness across variants
+        match self {
+            OTLPData::Logs(request) => {
+                "logs".hash(&mut hasher);
+                request.encode_to_vec().hash(&mut hasher);
+            }
+            OTLPData::Metrics(request) => {
+                "metrics".hash(&mut hasher);
+                request.encode_to_vec().hash(&mut hasher);
+            }
+            OTLPData::Traces(request) => {
+                "traces".hash(&mut hasher);
+                request.encode_to_vec().hash(&mut hasher);
+            }
+            OTLPData::Profiles(request) => {
+                "profiles".hash(&mut hasher);
+                request.encode_to_vec().hash(&mut hasher);
+            }
+        }
+
+        hasher.finish()
+    }
+
+    fn deadline(&self) -> Option<std::time::Instant> {
+        // OTLP messages don't have built-in deadlines
+        // Could be extended to extract deadline from headers/metadata in the future
```

Comment:
I believe what the otelarrow exporter does is the same as what gRPC does, fwiw, the `grpc-timeout` header. I'm not sure how OTLP-over-HTTP handles this, though.